File: 1-introduction/c/interp.c
#include <stdio.h>
#include "util.h"
#include "slp.h"
#include "table.h"
#include "interp.h"

typedef struct IntAndTable {
  int i;
  Table_ t;
} IntAndTable;

IntAndTable interpExp(A_exp e, Table_ t);

Table_ interpExpList(A_expList expList, Table_ t) {
  switch (expList->kind) {
    case A_pairExpList:
    {
      A_exp head = expList->u.pair.head;
      IntAndTable headResult = interpExp(head, t);
      t = headResult.t;
      A_expList tail = expList->u.pair.tail;
      t = interpExpList(tail, t);
      return t;
    }
    case A_lastExpList:
    {
      A_exp last = expList->u.last;
      IntAndTable lastResult = interpExp(last, t);
      return lastResult.t;
    }
  }
}

Table_ interpStm(A_stm s, Table_ t) {
  switch (s->kind) {
    case A_compoundStm:
      {
        t = interpStm(s->u.compound.stm1, t);
        t = interpStm(s->u.compound.stm2, t);
        return t;
      }
    case A_assignStm:
      {
        string key = s->u.assign.id;
        IntAndTable intAndTable = interpExp(s->u.assign.exp, t);
        return update(intAndTable.t, key, intAndTable.i);
      }
    case A_printStm:
      {
        A_expList expList= s->u.print.exps;
        return interpExpList(expList, t);
      }
    default:
      printf ("unknown stm kind %d\n", s->kind);
   }
}

IntAndTable interpExp(A_exp e, Table_ t) {
  IntAndTable intAndTable = {0, t};
  switch (e->kind) {
    case A_idExp:
      {
        IntAndTable result = {lookup(t, e->u.id), t};
        return result;
      }
    case A_numExp:
      {
        IntAndTable result = {e->u.num, t};
        return result;
      }
    case A_opExp:
      {
        A_exp left = e->u.op.left;
        A_binop oper = e->u.op.oper;
        A_exp right = e->u.op.right;
        IntAndTable leftResult = interpExp(left, t);
        int leftNum = leftResult.i;
        t = leftResult.t;
        IntAndTable rightResult = interpExp(right, t);
        int rightNum = rightResult.i;
        t = rightResult.t;
        int num = 0;
        switch (oper) {
          case A_plus:
            num = leftNum + rightNum;
            break;
          case A_minus:
            num = leftNum - rightNum;
            break;
          case A_times:
            num = leftNum * rightNum;
            break;
          case A_div:
            num = leftNum / rightNum;
            break;
        }
        IntAndTable result = {num, t};
        return result;
      }
    case A_eseqExp:
      {
        A_stm stm = e->u.eseq.stm;
        A_exp exp = e->u.eseq.exp;
        t = interpStm(stm, t);
        IntAndTable expResult = interpExp(exp, t);
        return expResult;
      }
  default:
    printf ("unknown exp kind %d\n", e->kind);
    exit(1);
    break;
  }
  return intAndTable;
}

File: 1-introduction/ts/src/interp.ts
import { Expression, ExpressionList, IdExpression, NumExpression, OpExpression, EseqExpression, ExpressionPair, ExpressionLast } from './types/expression';
import { Statement, AssignStatement, CompoundStatement, PrintStatement } from './types/statement';
type Table = Record<string, number>;

function newTable(id: string, val: number, tail: Table): Table {
  return {
    ...tail,
    [id]: val
  };
}

type ValAndTable = { val: number, table: Table };

export function interp(s: Statement, t: Table): Table {
  switch (s.kind) {
    case 'AssignStatement':
      const assign = s as AssignStatement;
      const result = interpExp(assign.expression, t);
      return newTable(assign.id, result.val, result.table);
    case 'CompoundStatement':
      const compound = s as CompoundStatement;
      const t1 = interp(compound.st1, t);
      const t2 = interp(compound.st2, t1);
      return t2;
    case 'PrintStatement':
      const print = s as PrintStatement;
      return interpExps(print.exps, t);
  }
}

function interpExp(exp: Expression, t: Table): ValAndTable {
  switch (exp.kind) {
    case 'IdExpression':
      const idExp = exp as IdExpression;
      return { val: t[idExp.id], table: t };
    case 'NumExpression':
      return { val: (exp as NumExpression).num, table: t };
    case 'OpExpression':
      return interpOpExp(exp as OpExpression, t);
    case 'EseqExpression':
      return interpEseqExp(exp as EseqExpression, t);

  }
}

function interpExps(exps: ExpressionList, t: Table): Table {
  switch (exps.kind) {
    case 'pair':
      const pair = exps as ExpressionPair;
      const res1 = interpExp(pair.head, t);
      return interpExps(pair.tail, res1.table);
    case 'last':
      const last = exps as ExpressionLast;
      const res2 = interpExp(last.last, t);
      return res2.table;
  }
}

function interpOpExp(opExp: OpExpression, t: Table): ValAndTable {
  const { val: left, table: t1 } = interpExp(opExp.left, t);
  const { val: right, table: t2 } = interpExp(opExp.right, t1);
  let val: number;
  switch (opExp.op) {
    case '*':
      val = left * right;
      break;
    case '+':
      val = left + right;
      break;
    case '-':
      val = left - right;
      break;
    case '/':
      val = left / right;
      break;
  }
  return { val, table: t2 };
}

function interpEseqExp(eseq: EseqExpression, t: Table): ValAndTable {
  t = interp(eseq.statement, t);
  return interpExp(eseq.expression, t);
}
